File: 1306-jump-game-iii/1306-jump-game-iii.cpp
class Solution {
public:
    bool bfs(vector<int>& arr, vector<int>& visited, queue<int> &q ,int indx){
        if(indx < 0 || indx >= arr.size() || visited[indx]) 
          return false;
        if(arr[indx] == 0) return true;
        q.push(indx);
        visited[indx] = 1;
        return false;
    }
    bool canReach(vector<int>& arr, int start) {
        if(arr[start] == 0 ) return true;
        vector<int> visited(arr.size(), 0);
        queue<int> q;
        q.push(start);
        visited[start] = 1;
        while(!q.empty()){
            int front = q.front();
            q.pop();
            if(bfs(arr,visited,q ,front + arr[front])||
            bfs(arr,visited, q ,front - arr[front])){
                return true;
            }
        }
        return false;
    }
};

File: 1306-jump-game-iii/1306-jump-game-iii.py
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        if arr[start] == 0: return True
        def bfs(visited,q,indx):
            if indx < 0 or indx >= len(arr) or visited[indx]:
                return False
            if arr[indx] == 0: 
                return True
            q.append(indx)
            visited[indx] = 1
            return False

        visited = [0]*len(arr)
        q = deque()
        q.append(start)
        visited[start] = 1
        while q:
            front = q.popleft()
            if bfs(visited,q,front + arr[front]) or bfs(visited,q,front - arr[front]):
                return True
       
        return False


        

        

        